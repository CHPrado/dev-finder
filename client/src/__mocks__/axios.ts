const axios = {
  get: jest.fn(() => Promise.resolve({})),
  post: jest.fn(() => Promise.resolve({})),
  create: () => axios,
  defaults: {
    adapter: {},
  },
};

export default axios;
