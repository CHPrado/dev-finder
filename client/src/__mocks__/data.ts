import { createMemoryHistory, createLocation } from "history";

import { DevProps, DevRouteProps } from "../interfaces";

const devData: DevProps = {
  username: "CHPrado",
  name: "Christian Prado",
  bio: "My bio",
  avatar_url: "avatar_url",
  location: "Lins - SP",
  followers: 1,
  following: 2,
  public_repos: 3,
  stars: 4,
  watchers: 5,
  forks: 6,
  languages: [
    {
      name: "JavaScript",
      count: 1,
      percent: "49.9",
    },
    {
      name: "Ruby",
      count: 1,
      percent: "51.1",
    },
  ],
  notes: "Notes about me",
};

const devRouteData: DevRouteProps = {
  history: createMemoryHistory(),
  location: createLocation({ state: { dev: devData } }),
  match: {
    isExact: false,
    path: "/dev",
    url: "/dev",
    params: {},
  },
};

export { devData, devRouteData };
