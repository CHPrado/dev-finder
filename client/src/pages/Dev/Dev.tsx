import React, { FC, useState } from "react";
import { Link } from "react-router-dom";

import {
  Box,
  Card,
  CircularProgress,
  IconButton,
  Tooltip,
  Typography,
} from "@material-ui/core";
import { CachedOutlined } from "@material-ui/icons";
import { LocationIcon } from "@primer/octicons-react";

import { Header, Snackbar } from "../../components";
import { SnackbarProps, DevRouteProps } from "../../interfaces";
import { githubApi } from "../../services";
import { Details } from "./Details";
import { Languages } from "./Languages";
import { Notes } from "./Notes";
import useStyles from "./styles";

const Dev: FC<DevRouteProps> = (props) => {
  const classes = useStyles();
  const [dev, setDev] = useState(props.location.state.dev);
  const [isLoading, setIsLoading] = useState(false);
  const [snackbar, setSnackbar] = useState<SnackbarProps>({});

  const handleRefresh = () => {
    setIsLoading(true);
    githubApi
      .getDev(dev.username)
      .then((resp) => {
        setDev(resp);
      })
      .catch(() => {
        setSnackbar({
          open: true,
          message:
            "There was an error trying to refresh the data. Please, try again.",
          type: "error",
          onClose: () => setSnackbar({ open: false }),
        });
      })
      .finally(() => setIsLoading(false));
  };

  return (
    <Box className={classes.root}>
      <Header />

      {isLoading ? (
        <Box className={classes.loadingBox}>
          <CircularProgress size="100px" />
        </Box>
      ) : (
        <Box className={classes.container}>
          <Card variant="outlined" className={classes.card}>
            <Box className={classes.titleBox}>
              <Link
                className={classes.link}
                to={(location) => ({
                  ...location,
                  pathname: `http://www.github.com/${dev.username}`,
                })}
                target="_blank"
              >
                <Typography className={classes.name}>
                  {`${dev.username}'s profile`}
                </Typography>
              </Link>

              <Box display="flex">
                <Tooltip title="Refresh">
                  <IconButton onClick={handleRefresh}>
                    <CachedOutlined />
                  </IconButton>
                </Tooltip>
              </Box>
            </Box>

            <Box className={classes.avatarBox}>
              <img
                src={dev.avatar_url}
                alt="avatar"
                className={classes.avatar}
              />

              <Box className={classes.infoBox}>
                <Typography className={classes.info}>{dev.name}</Typography>
                {dev.location && (
                  <Typography className={classes.info}>
                    <LocationIcon size={18} className={classes.icon} />
                    {dev.location}
                  </Typography>
                )}
                <Typography color="textSecondary" align="justify">
                  {dev.bio}
                </Typography>
              </Box>
            </Box>

            <Box className={classes.detailsBox}>
              <Details dev={dev} />
              <Languages languages={dev.languages} />
            </Box>

            <Notes
              dev={dev}
              setIsLoading={setIsLoading}
              setSnackbar={setSnackbar}
              setDev={setDev}
            />
          </Card>
        </Box>
      )}

      <Snackbar {...snackbar} />
    </Box>
  );
};

export default Dev;
