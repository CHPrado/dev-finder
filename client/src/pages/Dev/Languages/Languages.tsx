import React, { FC } from "react";

import { Box, Typography } from "@material-ui/core";

import { DevProps } from "../../../interfaces";
import useStyles from "./styles";

interface Props {
  languages: DevProps["languages"];
}

const Languages: FC<Props> = ({ languages }) => {
  const classes = useStyles();

  return (
    <Box mt={{ xs: 4, sm: 0 }}>
      {languages.map((language) => (
        <Box key={language.name} className={classes.languageBox}>
          <Box width="100%">
            <Typography className={classes.language}>
              {language.name}
            </Typography>
          </Box>
          <Box width="100%">
            <Typography className={classes.percent}>
              {language.percent}%
            </Typography>
          </Box>
        </Box>
      ))}
    </Box>
  );
};

export default Languages;
