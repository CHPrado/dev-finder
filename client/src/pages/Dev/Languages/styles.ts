import { Theme, makeStyles } from "@material-ui/core/styles";

export default makeStyles((theme: Theme) => ({
  languageBox: {
    display: "flex",
    justifyContent: "space-between",
    [theme.breakpoints.down("xs")]: {
      justifyContent: "center",
    },
  },
  language: {
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    whiteSpace: "nowrap",
    [theme.breakpoints.down("xs")]: {
      textAlign: "right",
    },
  },
  percent: {
    fontWeight: theme.typography.fontWeightBold,
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    textAlign: "right",
    [theme.breakpoints.down("xs")]: {
      textAlign: "left",
    },
  },
}));
