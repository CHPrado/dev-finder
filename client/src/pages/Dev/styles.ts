import { Theme, makeStyles } from "@material-ui/core/styles";

const flexCenter = {
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
};

export default makeStyles((theme: Theme) => ({
  root: {
    display: "flex",
    flexDirection: "column",
    height: "100%",
  },
  loadingBox: {
    height: "100%",
    ...flexCenter,
  },
  container: {
    ...flexCenter,
    position: "absolute",
    top: "100px",
    width: "100%",
  },
  card: {
    ...flexCenter,
    flexDirection: "column",
    margin: theme.spacing(5),
    padding: theme.spacing(5),
    maxWidth: "500px",
    [theme.breakpoints.down("xs")]: {
      width: "100%",
    },
  },
  titleBox: {
    width: "100%",
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    marginBottom: theme.spacing(4),
  },
  name: {
    color: theme.palette.primary.dark,
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightBold,
    whiteSpace: "nowrap",
  },
  avatarBox: {
    display: "flex",
    alignItems: "flex-start",
    justifyContent: "center",
    [theme.breakpoints.down("xs")]: {
      flexDirection: "column",
      alignItems: "center",
    },
  },
  avatar: {
    borderRadius: "50%",
    height: "150px",
    marginRight: theme.spacing(4),
    [theme.breakpoints.down("xs")]: {
      height: "200px",
      marginBottom: theme.spacing(4),
    },
  },
  infoBox: {
    marginBottom: theme.spacing(4),
    display: "flex",
    flexDirection: "column",
    width: "100%",
  },
  info: {
    color: theme.palette.text.secondary,
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightBold,
    display: "flex",
    alignItems: "center",
  },
  detailsBox: {
    width: "100%",
    display: "flex",
    justifyContent: "center",
    margin: theme.spacing(4),
    [theme.breakpoints.down("xs")]: {
      flexDirection: "column",
    },
  },
  icon: {
    marginRight: theme.spacing(2),
  },
  link: {
    textDecoration: "none",
    "& :hover": {
      color: theme.palette.primary.main,
      textDecoration: "underline",
    },
  },
}));
