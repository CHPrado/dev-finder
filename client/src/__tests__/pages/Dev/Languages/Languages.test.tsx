import React from "react";

import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { devData } from "../../../../__mocks__/data";
import { Languages } from "../../../../pages/Dev/Languages";

describe("rendering component", () => {
  it("renders Notes component without crashing", () => {
    const wrapper = shallow(<Languages languages={devData.languages} />);

    expect(wrapper.contains(devData.languages[0].name)).toEqual(true);
    expect(wrapper.contains(`${devData.languages[0].percent}`)).toEqual(true);
    expect(wrapper.contains(devData.languages[1].name)).toEqual(true);
    expect(wrapper.contains(`${devData.languages[1].percent}`)).toEqual(true);
  });
});

describe("passing props", () => {
  it("laguages props data equals to dev languages data", () => {
    const wrapper = mount(<Languages languages={devData.languages} />);

    expect(wrapper.props().languages).toEqual(devData.languages);
  });
});

describe("snapshots", () => {
  it("Languages snapshots", () => {
    const tree = shallow(<Languages languages={devData.languages} />);

    expect(toJson(tree)).toMatchSnapshot();
  });
});
