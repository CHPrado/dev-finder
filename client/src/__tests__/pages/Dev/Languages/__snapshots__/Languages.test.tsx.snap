// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots Languages snapshots 1`] = `
<Styled(MuiBox)
  mt={
    Object {
      "sm": 0,
      "xs": 4,
    }
  }
>
  <Styled(MuiBox)
    className="makeStyles-languageBox-1"
    key="JavaScript"
  >
    <Styled(MuiBox)
      width="100%"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-language-2"
      >
        JavaScript
      </WithStyles(ForwardRef(Typography))>
    </Styled(MuiBox)>
    <Styled(MuiBox)
      width="100%"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-percent-3"
      >
        49.9
        %
      </WithStyles(ForwardRef(Typography))>
    </Styled(MuiBox)>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    className="makeStyles-languageBox-1"
    key="Ruby"
  >
    <Styled(MuiBox)
      width="100%"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-language-2"
      >
        Ruby
      </WithStyles(ForwardRef(Typography))>
    </Styled(MuiBox)>
    <Styled(MuiBox)
      width="100%"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-percent-3"
      >
        51.1
        %
      </WithStyles(ForwardRef(Typography))>
    </Styled(MuiBox)>
  </Styled(MuiBox)>
</Styled(MuiBox)>
`;
